libnavigation-base/src/main/java/com/mapbox/navigation/base/routerefresh/RoutingOptionsProvider.kt

```diff
@@ -0,0 +1,25 @@
+package com.mapbox.navigation.base.routerefresh
+
+/**
+ * Provider for processing routing events
+ */
+open class RoutingOptionsProvider {
+
+    private val routeRefreshAdapter: RouteRefreshAdapter by lazy {
+        MapboxRouteRefreshAdapter()
+    }
+
+    /**
+     * Defines adapter that will be used by SDK on off route event.
+     * By default uses [MapboxRouteRefreshAdapter]
+     */
+    open fun provideOffRouteRefreshAdapter(): RouteRefreshAdapter = routeRefreshAdapter
```

Comment:
We should flatten this. Delete MapboxRouteRefreshAdapter in favor of only RoutingOptionsProvider.

As in, RoutingOptionsProvider is the interface
MapboxRoutingOptionsProvider is the default implementation (remove the adapters)

Reply:
@kmadsen think that's not scalable. `MapboxNavigation` extends constructor for that `RoutingOptionsProvider`, but if we'd like to send one more, as instance for Faster route, we have to extend `MapboxNavigation` constructor again, make sense? Idea of this provider is that we might add here as much Adapters as we'd like and don't change `MapboxNavigation` API

Reply:
RoutingOptionsProvider as is providing an Adapter not RoutingOptions. So the comment is to have this be an interface, and provide the routing options

``` kotlin
interface RouteOptionsProvider {
    fun newRouteOptions(
        routeOptions: RouteOptions?,
        routeProgress: RouteProgress?,
        location: Location?
    ): RouteOptions?
}
```

and then the implementation

``` kotlin
class MapboxRouteOptionsProvider {
    override newRouteOptions(...) {
        // the code that is in RouteRefreshAdapter right now
    }
}
```

> MapboxNavigation extends constructor for that RoutingOptionsProvider, but if we'd like to send one more, as instance for Faster route, we have to extend MapboxNavigation constructor again, make sense?

Excuse my rephrase, only doing it to ensure I am understanding. I am reading your comment as "if we'd like one implementation of RoutingOptionsProvider for faster route, and a different implementation for route refresh; it won't scale because we have to extend the MapboxNavigation constructor - the classes would be in conflict"

Because RoutingOptionProvider would be an interface, it would work the same as with the RouteRefreshAdapter as an interface right?

Also we don't want to provide the RoutingOptionProvider through the constructor, we're working to remove optional features from the constructor https://github.com/mapbox/mapbox-navigation-android/issues/3187

Reply:
@kmadsen from this perspective it makes sense. I've got your think from scratch regarding Provider, but didn't get how to set it to `MapboxNavigation`. Now it's clear 👍 . From my perspective it's not so flexible, moreover, we expose default implementation, that is not cool because user doesn't need it.  
Could you 👀 again?

Reply:
> From my perspective it's not so flexible, moreover, we expose default implementation, that is not cool because user doesn't need it.
Could you 👀 again?

@RingerJK Do we need to expose the interface either? If we are exposing the interface, the default implementation serves as an example. Looks like we're [adding the ability to override it here](https://github.com/mapbox/mapbox-navigation-android/pull/3184/files#diff-e3ca7b557efaeed833ca6f8f76dc2d03R556). If that's our intention, looks good

This is related to another topic @Guardiola31337 is raising here though.  https://github.com/mapbox/mapbox-navigation-android/pull/3184#discussion_r442361150. 
Let's move the topic for how/where to keep instances of the class to that thread

Reply:
@kmadsen I've separated they to have ability override them independently but I'm open to discussion. If you remember previous implementation with Adapter/Provider, it was a good start point to not make n-methods(like `fun setRouteOptionProviderOfOLOLO`) into MapboxNavigation but only one(like `fun setProviderOfAdaptersRouteOptions` or smth like this)